hikari/interactions.py

```diff
@@ -0,0 +1,190 @@
+# -*- coding: utf-8 -*-
+# cython: language_level=3
+# Copyright (c) 2020 Nekokatt
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""Application and entities that are used to describe interactions on Discord."""
+
+from __future__ import annotations
+
+__all__: typing.List[str] = [
+    "OptionType",
+    "OptionValueT",
+    "CommandChoice",
+    "CommandOption",
+    "Command",
+    "InteractionType",
+    "CommandInteractionData",
+    "CommandInteractionOption",
+    "CommandInteraction",
+]
+
+import typing
+
+import attr
+
+from hikari import guilds
+from hikari import permissions as permissions_
+from hikari import snowflakes
+from hikari import traits
+from hikari.internal import attr_extensions
+from hikari.internal import enums
+
+if typing.TYPE_CHECKING:
+    from hikari import channels
+
+
+RawOptionValueT = typing.Union[str, int, bool]
+OptionValueT = typing.Union[str, int, bool, snowflakes.Snowflake]
+
+
+@typing.final
+class OptionType(int, enums.Enum):
+    SUB_COMMAND = 1
+    SUB_COMMAND_GROUP = 2
+    STRING = 3
+    STR = STRING
+    INTEGER = 4
+    INT = INTEGER
+    BOOLEAN = 5
+    BOOL = BOOLEAN
+    USER = 6
+    CHANNEL = 7
+    ROLE = 8
+
+
+@typing.final
+class InteractionType(int, enums.Enum):
+    # PING isn't here as it should be handled as internal detail of the REST
+    # server rather than as a part of the public interface.
```

Comment:
Document it as being a internal, but have it in this enum

Reply:
And then presumably also make two union types that look like
```py
InteractionTypeT = typing.Literal[InteractionType[APPLICATION_COMMAND]]
```
and
```py
InteractionResponseTypeT = typing.Union[
    typing.Literal[InteractionResponseType.ACKNOWLEDGE],
    typing.Literal[InteractionResponseType.CHANNEL_MESSAGE],
    typing.Literal[InteractionResponseType.CHANNEL_MESSAGE_WITH_SOURCE],
    typing.Literal[InteractionResponseType.ACKNOWLEDGE_WITH_SOURCE],
]
```
to match that behaviour?